lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,560 @@
+  return isCharType(subtagRef, 1, 8, &isASCIILetterOrDigit);
+}
+bool isOtherExtension(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{2,8})+;
+  return isCharType(subtagRef, 2, 8, &isASCIILetterOrDigit);
+}
+}
+
+class LanguageTagParser {
+ public:
+  LanguageTagParser(const std::u16string &localeId);
+  LanguageTagParser();
+  ~LanguageTagParser();
+
+  // public function declaration
+  bool parseUnicodeLocaleId();
+  ParsedLocaleIdentifier getParsedLocaleId();
+  std::u16string toString();
+
+ private:
+  // private function declaration
+  bool parseUnicodeLanguageId(bool transformedExtensionId);
+  bool addVariantSubtag(bool transformedExtensionId);
+  bool parseExtensions();
+  bool parseUnicodeExtension();
+  bool parseTransformedExtension();
+  bool parseOtherExtension(char16_t singleton);
+  bool parsePUExtension();
+  // tokenizer functions
+  std::u16string getCurrentSubtag();
+  llvh::ArrayRef<char16_t> getCurrentSubtagRef();
+  bool hasMoreSubtags();
+  bool nextSubtag();
+  
+  ParsedLocaleIdentifier parsedLocaleIdentifier;
+  std::u16string localeId_;
+  std::u16string originalLocaleId_;
+  int subtagStart_;
+  int subtagEnd_;
+};
+
+LanguageTagParser::LanguageTagParser(const std::u16string &localeId) {
+  localeId_ = localeId;
+  toASCIILowerCase(localeId_);
+  originalLocaleId_ = localeId;
+  subtagStart_ = 0;
+  subtagEnd_ = -1;
+}
+LanguageTagParser::~LanguageTagParser() = default;
+
+ParsedLocaleIdentifier LanguageTagParser::getParsedLocaleId(){
+  return parsedLocaleIdentifier;
+}
+
+bool LanguageTagParser::parseUnicodeLocaleId() {
+  if (!parseUnicodeLanguageId(false))  {
+    return false;
+  }
+  if (!hasMoreSubtags()) {
+    return true;
+  }
+  if (!parseExtensions()) {
+    return false;
+  }
+  
+  return !hasMoreSubtags();
+}
+
+bool LanguageTagParser::parseUnicodeLanguageId(bool transformedExtensionId) {
+  if (!transformedExtensionId) {
+    if (!nextSubtag() || !isUnicodeLanguageSubtag(getCurrentSubtagRef())) {
+      return false;
+    }
+  }
+ 
+  ParsedLanguageIdentifier *languageId;
+  
+  if (transformedExtensionId) {
+    languageId = &parsedLocaleIdentifier.transformedLanguageIdentifier;
+  } else {
+    languageId = &parsedLocaleIdentifier.languageIdentifier;
+  }
+  
+  languageId->languageSubtag = getCurrentSubtag();
+  
+  if (!nextSubtag()) {
+    return true;
+  }
+  
+  if (isUnicodeScriptSubtag(getCurrentSubtagRef())) {
+    languageId->scriptSubtag = getCurrentSubtag();
+    if (!transformedExtensionId) {
+      toASCIITitleCase(languageId->scriptSubtag);
+    }
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+
+  if (isUnicodeRegionSubtag(getCurrentSubtagRef())) {
+    languageId->regionSubtag = getCurrentSubtag();
+    if (!transformedExtensionId) {
+      toASCIIUpperCase(languageId->regionSubtag);
+    }
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+  
+  while (true) {
```

Comment:
You may want to consider making this a do..while loop. Loosely, it could then be restructured as:
```
do { 
  // body
} while (nextSubTag);

return true;
```